usaspending_api/references/v2/views/naics.py

```diff
@@ -0,0 +1,178 @@
+import logging
+from collections import OrderedDict
+
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from django.db.models.functions import Length
+
+from usaspending_api.common.cache_decorator import cache_response
+from usaspending_api.common.validator.tinyshield import TinyShield
+from usaspending_api.references.models import NAICS
+
+logger = logging.getLogger("console")
+
+
+class NAICSViewSet(APIView):
+    """
+    Return a list of NAICS or a filtered list of NAICS
+    """
+
+    endpoint_doc = "usaspending_api/api_contracts/contracts/v2/references/naics.md"
+
+    def _parse_and_validate_request(self, requested_naics: str, request_data) -> dict:
+        naics_filter = request_data.get("filter")
+        data = {"code": requested_naics, "description": naics_filter}
+        models = [
+            {"key": "code", "name": "code", "type": "integer", "allow_nulls": True, "optional": True},
+            {
+                "key": "description",
+                "name": "description",
+                "type": "text",
+                "text_type": "search",
+                "default": None,
+                "optional": True,
+                "allow_nulls": True,
+            },
+        ]
+        validated = TinyShield(models).block(data)
+        return validated
+
+    def _fetch_children(self, naics_code, naics_filter=None) -> list:
+        results = []
+        if naics_filter:
+            naics = NAICS.objects.annotate(text_len=Length("code")).filter(
+                **naics_filter, code__startswith=naics_code, text_len=len(naics_code) + 2
+            )
+        else:
+            naics = NAICS.objects.annotate(text_len=Length("code")).filter(
+                code__startswith=naics_code, text_len=len(naics_code) + 2
+            )
+        for naic in naics:
+            if len(naic.code) < 6:
+                result = OrderedDict()
+                result["naics"] = naic.code
+                result["naics_description"] = naic.description
+                result["count"] = (
+                    NAICS.objects.annotate(text_len=Length("code"))
+                    .filter(code__startswith=naic.code, text_len=6)
+                    .count()
+                )
+            else:
+                result = OrderedDict()
+                result["naics"] = naic.code
+                result["naics_description"] = naic.description
+                result["count"] = 0
```

Comment:
We might save some time by skipping the search sometimes, but do these two cases need to be separated? What if in the event of leaf nodes we just searched for children anyway and returned the zero results?
That said, it looks like there might be some special logic here. If I fetch the children of a naics of length 2, it will find children at length 4, then set the count of each result as the total children of length 6. If I put in a naics of length 4, it will find children of length six and set all their counts to 0, even though they each have one naic at length six starting with the code (themselves). Put in a naics of length 6, and it returns no children at all (reasonably). Is this how we want this to work?

Reply:
Yes. This is the behavior specifically requested in the ticket. 
https://federal-spending-transparency.atlassian.net/browse/DEV-3527
> In all cases, count is the number of Tier 3 nodes that fall under the NAICS requested.  Tier 3 nodes will always have a count of 0

Reply:
Hmm.  @BrianZito makes a good point here.  Need to check with @jonhill13.  Should Tier 3 return a count of 1 instead of 0?  I was thinking of it in terms of "count of the node's Tier 3 ancestors", but perhaps some front end math could be genericized if it always returned 1 instead?

Reply:
Confirmed with  @jonhill13, we are changing the Tier 3s to return 1 instead of 0 for count.